### PR TITLE
modify qa-trainer

### PR DESCRIPTION
**Pull request by zhangfanTJU**
I fixed the evaluation failure for the TrainerQA-based script [`examples/pytorch/question-answering/run_qa.py`](https://github.com/huggingface/transformers/blob/master/examples/pytorch/question-answering/run_qa.py) when using distributed training, which has been partially fixed in  #11746.

```
Traceback (most recent call last):
  File "run_qa.py", line 543, in <module>  
    main()
  File "run_qa.py", line 509, in main
    metrics = trainer.evaluate()
  File "trainer_qa.py", line 44, in evaluate
    ignore_keys=ignore_keys,
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/transformers/trainer.py", line 2162, in evaluation_loop
    logits = self._nested_gather(logits)
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/transformers/trainer.py", line 2252, in _nested_gather
    tensors = distributed_concat(tensors)
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/transformers/trainer_pt_utils.py", line 154, in distributed_concat
    return type(tensor)(distributed_concat(t, num_total_examples) for t in tensor)
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/transformers/trainer_pt_utils.py", line 154, in <genexpr>
    return type(tensor)(distributed_concat(t, num_total_examples) for t in tensor)
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/transformers/trainer_pt_utils.py", line 156, in distributed_concat
    dist.all_gather(output_tensors, tensor)
  File "/home/x/anaconda3/envs/torch18/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1863, in all_gather
    work = default_pg.allgather([tensor_list], [tensor])
RuntimeError: Tensors must be non-overlapping and dense
```

This failure is similar to the the previous commit (https://github.com/huggingface/transformers/pull/404/commits/fda2f623953bfe2290cd65429eb008f02ebdb152), but it also happens in pytorch 1.8 now.

Meanwhile, I added **Evaluation** and **Prediction** log for the script [`run_qa_no_trainer.py`](https://github.com/huggingface/transformers/blob/master/examples/pytorch/question-answering/run_qa_no_trainer.py) according to the TrainerQA.

Thanks! @sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for tackling this! Could you also ad that change to the model templates?

**Review by LysandreJik — APPROVED** [review]
Cool!

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 1]
```diff
@@ -1152,8 +1152,8 @@ def __call__(
 
         logits = self.qa_outputs(hidden_states)
         start_logits, end_logits = logits.split(self.config.num_labels, axis=-1)
-        start_logits = start_logits.squeeze(-1)
-        end_logits = end_logits.squeeze(-1)
+        start_logits = start_logits.squeeze(-1).contiguous()
+        end_logits = end_logits.squeeze(-1).contiguous()
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_flax_bert.py:1156]: I don't think this should be here, this is a Flax model

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 2]
```diff
@@ -1031,8 +1031,8 @@ def __call__(
         hidden_states = outputs[0]
         logits = self.qa_outputs(hidden_states)
         start_logits, end_logits = logits.split(self.config.num_labels, axis=-1)
-        start_logits = start_logits.squeeze(-1)
-        end_logits = end_logits.squeeze(-1)
+        start_logits = start_logits.squeeze(-1).contiguous()
+        end_logits = end_logits.squeeze(-1).contiguous()
```
Comment (LysandreJik) [inline src/transformers/models/electra/modeling_flax_electra.py:1035]: I don't think this should be here, this is a Flax model

File: src/transformers/models/roberta/modeling_flax_roberta.py [hunk 3]
```diff
@@ -977,8 +977,8 @@ def __call__(
 
         logits = self.qa_outputs(hidden_states)
         start_logits, end_logits = logits.split(self.config.num_labels, axis=-1)
-        start_logits = start_logits.squeeze(-1)
-        end_logits = end_logits.squeeze(-1)
+        start_logits = start_logits.squeeze(-1).contiguous()
+        end_logits = end_logits.squeeze(-1).contiguous()
```
Comment (LysandreJik) [inline src/transformers/models/roberta/modeling_flax_roberta.py:981]: I don't think this should be here, this is a Flax model

**Review by LysandreJik — CHANGES_REQUESTED** [review]

File: src/transformers/models/roberta/modeling_flax_roberta.py (hunk 3 above)
Comment (zhangfanTJU) [reply src/transformers/models/roberta/modeling_flax_roberta.py]: I'm really sorry that I don't know much about Flax model, but I have made some modifications. Many thanks!

**Comment by zhangfanTJU** [comment]
I encountered an unexpected error in ci, could you help me to finish this pr? @LysandreJik

**Comment by LysandreJik** [comment]
Seems like a CircleCI failure, I just relaunched all tests.

**Comment by zhangfanTJU** [comment]
the CircleCI encountered a HTTPError, what should I do?

**Review by LysandreJik — APPROVED** [review]
LGTM!

**Pull request merged**